File: python/nueva/ace_step.py
"""
ACE-Step 1.5 Integration for Nueva

This module provides the interface to the ACE-Step 1.5 music generation model.
It handles model loading, inference, and audio processing.

ACE-Step 1.5 capabilities per spec §5.3:
- text_to_music: Generate music from text description
- cover: Create covers in different styles
- repaint: Modify specific aspects of audio
- style_change: Change the overall style
- track_extraction: Extract stems/tracks
- layering: Add layers to existing audio
- completion: Complete partial audio
"""

import os
import sys
import json
import subprocess
import tempfile
from pathlib import Path
from typing import Any, Optional
import shutil


class AceStepProcessor:
    """ACE-Step 1.5 processor for Nueva."""

    # ACE-Step installation path (configurable via env var)
    ACE_STEP_PATH = os.environ.get(
        "NUEVA_ACE_STEP_PATH",
        str(Path.home() / "ACE-Step-1.5")
    )

    # Default model parameters
    DEFAULT_PARAMS = {
        "mode": "transform",
        "preserve_melody": True,
        "intensity": 0.7,
        "inference_steps": 8,
        "guidance_scale": 3.0,
    }

    # Mode mappings
    MODES = {
        "transform": "Transform audio based on prompt",
        "repaint": "Repaint/modify specific aspects",
        "cover": "Create a cover version",
        "extract": "Extract stems/tracks",
        "layer": "Add layers to audio",
        "complete": "Complete partial audio",
    }

    def __init__(self):
        self._model_loaded = False
        self._pipeline = None
        self._check_installation()

    def _check_installation(self):
        """Check if ACE-Step is installed."""
        ace_path = Path(self.ACE_STEP_PATH)
        self._installed = ace_path.exists() and (ace_path / "acestep").exists()

        if not self._installed:
            print(
                f"Warning: ACE-Step not found at {self.ACE_STEP_PATH}. "
                f"Install with: git clone https://github.com/ACE-Step/ACE-Step-1.5.git",
                file=sys.stderr
            )

    def is_available(self) -> bool:
        """Check if the model is available for use."""
        return self._installed

    def get_info(self) -> dict:
        """Get model information."""
        return {
            "id": "ace-step",
            "name": "ACE-Step 1.5",
            "version": "1.5",
            "description": "Full music transformation via Hybrid Reasoning-Diffusion",
            "capabilities": list(self.MODES.keys()),
            "use_when": [
                "Dramatic transformation",
                "Genre change",
                "Cover generation",
                "Reimagine as X",
                "Style transfer",
            ],
            "limitations": [
                "Takes several seconds to process",
                "Non-deterministic results",
                "Requires GPU for best performance",
            ],
            "known_artifacts": [
                "Vocal intelligibility loss on complex lyrics",
                "Tempo drift on pieces >5 minutes",
                "Transient softening on aggressive percussion",
            ],
            "vram_requirement_gb": 4.0,
            "inference_time": "1-30 seconds depending on GPU",
            "installed": self._installed,
            "install_path": self.ACE_STEP_PATH,
            "supported_params": [
                {
                    "name": "mode",
                    "type": "enum",
                    "options": list(self.MODES.keys()),
                    "default": "transform",
                    "description": "Processing mode",
                },
                {
                    "name": "prompt",
                    "type": "string",
                    "required": True,
                    "description": "Text description of desired output",
                },
                {
                    "name": "preserve_melody",
                    "type": "bool",
                    "default": True,
                    "description": "Whether to preserve the original melody",
                },
                {
                    "name": "intensity",
                    "type": "float",
                    "min": 0.0,
                    "max": 1.0,
                    "default": 0.7,
                    "description": "Transformation intensity",
                },
                {
                    "name": "inference_steps",
                    "type": "int",
                    "min": 4,
                    "max": 50,
                    "default": 8,
                    "description": "Number of diffusion steps (more = better quality, slower)",
                },
                {
                    "name": "guidance_scale",
                    "type": "float",
                    "min": 1.0,
                    "max": 10.0,
                    "default": 3.0,
                    "description": "How closely to follow the prompt",
                },
            ],
        }

    def _load_model(self):
        """Load the ACE-Step model (lazy loading)."""
        if self._model_loaded:
            return

        if not self._installed:
            raise RuntimeError(
                f"ACE-Step not installed. Please install from: "
                f"https://github.com/ACE-Step/ACE-Step-1.5"
            )

        # Import ACE-Step modules
        sys.path.insert(0, self.ACE_STEP_PATH)

        try:
            # Try to import the handler
            from acestep.acestep_v15_pipeline import AceStepHandler

            # Initialize handler (this loads the model)
            self._pipeline = AceStepHandler()
            self._model_loaded = True
            print("ACE-Step 1.5 model loaded successfully", file=sys.stderr)

        except ImportError as e:
            raise RuntimeError(f"Failed to import ACE-Step: {e}")
        except Exception as e:
            raise RuntimeError(f"Failed to load ACE-Step model: {e}")

    def process(
        self,
        input_path: Path,
        output_path: Path,
        prompt: Optional[str] = None,
        params: Optional[dict] = None,
    ) -> dict:
        """
        Process audio through ACE-Step.

        Args:
            input_path: Path to input audio file
            output_path: Path where output should be written
            prompt: Text description of desired transformation
            params: Additional parameters (mode, intensity, etc.)

        Returns:
            dict with processing results
        """
        if not self._installed:
            return self._process_via_api(input_path, output_path, prompt, params)

        # Merge with defaults
        params = {**self.DEFAULT_PARAMS, **(params or {})}

        if not prompt:
            prompt = params.get("prompt", "transform the audio")

        # Load model if needed
        self._load_model()

        mode = params.get("mode", "transform")
        intensity = params.get("intensity", 0.7)
        preserve_melody = params.get("preserve_melody", True)
        inference_steps = params.get("inference_steps", 8)
        guidance_scale = params.get("guidance_scale", 3.0)

        # Map our modes to ACE-Step task_types
        task_type_map = {
            "transform": "repaint",
            "repaint": "repaint",
            "cover": "cover",
            "extract": "repaint",
            "layer": "repaint",
            "complete": "repaint",
        }
        task_type = task_type_map.get(mode, "repaint")

        try:
            # Initialize service if needed
            if not hasattr(self._pipeline, '_initialized'):
                self._pipeline.initialize_service()
                self._pipeline._initialized = True

            # Run ACE-Step processing
            result = self._pipeline.generate_music(
                captions=prompt,
                lyrics="",
                src_audio=str(input_path),
                task_type=task_type,
                inference_steps=inference_steps,
                guidance_scale=guidance_scale,
                audio_cover_strength=intensity,
            )

            # Save output audio
            if result and "audio" in result:
                import soundfile as sf
                sf.write(str(output_path), result["audio"], result.get("sample_rate", 44100))

            # Determine intentional artifacts based on mode and prompt
            artifacts = self._detect_artifacts(mode, prompt, params)

            return {
                "success": True,
                "message": f"ACE-Step {mode}: '{prompt}' at {intensity*100:.0f}% intensity",
                "reasoning": f"Applied {mode} transformation using ACE-Step 1.5",
                "intentional_artifacts": artifacts,
                "metadata": {
                    "mode": mode,
                    "prompt": prompt,
                    "intensity": intensity,
                    "preserve_melody": preserve_melody,
                    "inference_steps": inference_steps,
                },
            }

        except Exception as e:
            raise RuntimeError(f"ACE-Step processing failed: {e}")

    def _process_via_api(
        self,
        input_path: Path,
        output_path: Path,
        prompt: Optional[str],
        params: Optional[dict],
    ) -> dict:
        """
        Process via ACE-Step API server if model not loaded locally.

        This allows running ACE-Step as a separate service.
        """
        import urllib.request
        import urllib.error

        api_url = os.environ.get("NUEVA_ACE_STEP_API", "http://localhost:8001")

        params = {**self.DEFAULT_PARAMS, **(params or {})}
        prompt = prompt or params.get("prompt", "transform the audio")

        # Prepare API request
        request_data = {
            "audio_path": str(input_path),
            "prompt": prompt,
            "output_path": str(output_path),
            "mode": params.get("mode", "transform"),
            "strength": params.get("intensity", 0.7),
            "preserve_melody": params.get("preserve_melody", True),
            "num_inference_steps": params.get("inference_steps", 8),
            "guidance_scale": params.get("guidance_scale", 3.0),
        }

        try:
            req = urllib.request.Request(
                f"{api_url}/process",
                data=json.dumps(request_data).encode("utf-8"),
                headers={"Content-Type": "application/json"},
                method="POST",
            )

            with urllib.request.urlopen(req, timeout=600) as response:
                result = json.loads(response.read().decode("utf-8"))

            if not result.get("success", False):
                raise RuntimeError(result.get("error", "Unknown API error"))

            mode = params.get("mode", "transform")
            artifacts = self._detect_artifacts(mode, prompt, params)

            return {
                "success": True,
                "message": f"ACE-Step {mode} (via API): '{prompt}'",
                "reasoning": "Processed via ACE-Step API server",
                "intentional_artifacts": artifacts,
                "metadata": result.get("metadata", {}),
            }

        except urllib.error.URLError as e:
            raise RuntimeError(
                f"Cannot connect to ACE-Step API at {api_url}. "
                f"Start the API server with: cd {self.ACE_STEP_PATH} && uv run acestep-api"
            )

    def _detect_artifacts(self, mode: str, prompt: str, params: dict) -> list:
        """Detect intentional artifacts based on mode and prompt."""
        artifacts = []
        prompt_lower = prompt.lower()

        # Mode-specific artifacts
        if mode == "cover":
            artifacts.append("different_timbre")
            artifacts.append("style_variation")

        if mode == "repaint":
            artifacts.append("localized_changes")

        # Prompt-based artifacts
        if "vintage" in prompt_lower or "retro" in prompt_lower:
            artifacts.append("intentional_coloration")
            artifacts.append("frequency_rolloff")

        if "lo-fi" in prompt_lower or "lofi" in prompt_lower:
            artifacts.append("bitcrushing")
            artifacts.append("noise")
            artifacts.append("sample_rate_artifacts")

        if "vinyl" in prompt_lower:
            artifacts.append("high_frequency_noise")
            artifacts.append("subtle_crackle")

        if "tape" in prompt_lower:
            artifacts.append("subtle_hiss")
            artifacts.append("saturation")

        if "8-bit" in prompt_lower or "chiptune" in prompt_lower:
            artifacts.append("quantization")
            artifacts.append("limited_polyphony")

        return artifacts


def main():
    """CLI entry point for ACE-Step processor."""
    import argparse

    parser = argparse.ArgumentParser(description="ACE-Step 1.5 processor for Nueva")
    parser.add_argument("input", help="Input audio file")
    parser.add_argument("output", help="Output audio file")
    parser.add_argument("-p", "--prompt", required=True, help="Transformation prompt")
    parser.add_argument(
        "-m", "--mode",
        choices=["transform", "repaint", "cover", "extract", "layer", "complete"],
        default="transform",
        help="Processing mode"
    )
    parser.add_argument(
        "-i", "--intensity",
        type=float,
        default=0.7,
        help="Transformation intensity (0.0-1.0)"
    )
    parser.add_argument(
        "--preserve-melody",
        action="store_true",
        default=True,
        help="Preserve original melody"
    )
    parser.add_argument(
        "--steps",
        type=int,
        default=8,
        help="Inference steps (4-50)"
    )
    parser.add_argument(
        "--info",
        action="store_true",
        help="Print model info and exit"
    )

    args = parser.parse_args()

    processor = AceStepProcessor()

    if args.info:
        info = processor.get_info()
        print(json.dumps(info, indent=2))
        return

    if not processor.is_available():
        print(
            f"Error: ACE-Step not installed at {processor.ACE_STEP_PATH}",
            file=sys.stderr
        )
        print(
            "Install with: git clone https://github.com/ACE-Step/ACE-Step-1.5.git",
            file=sys.stderr
        )
        sys.exit(1)

    params = {
        "mode": args.mode,
        "intensity": args.intensity,
        "preserve_melody": args.preserve_melody,
        "inference_steps": args.steps,
    }

    try:
        result = processor.process(
            input_path=Path(args.input),
            output_path=Path(args.output),
            prompt=args.prompt,
            params=params,
        )
        print(json.dumps(result, indent=2))
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/neural/acestep.rs
//! ACE-Step neural model implementation
//!
//! Real implementation that communicates with the Nueva AI Bridge,
//! which in turn talks to ACE-Step API.
//!
//! Implements the NeuralModel trait for ACE-Step 1.5.

use crate::error::{NuevaError, Result};
use crate::neural::gpu::{can_run_ace_step, GpuInfo, QuantizationLevel};
use crate::neural::model::{
    NeuralModel, NeuralModelInfo, NeuralModelParams, ParamSpec, ParamType, ProcessingResult,
};

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;
use std::path::Path;
use std::time::Instant;

/// ACE-Step processing modes
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AceStepMode {
    /// Generate music from text prompt (text2music)
    Transform,
    /// Create a cover version - change style, preserve structure
    Cover,
    /// Repaint/modify specific audio regions
    Repaint,
    /// Source separation / extraction
    Extract,
    /// Add/remove instrument layers (lego mode)
    Layer,
    /// Add accompaniment / complete
    Complete,
}

impl AceStepMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Transform => "transform",
            Self::Cover => "cover",
            Self::Repaint => "repaint",
            Self::Extract => "extract",
            Self::Layer => "layer",
            Self::Complete => "complete",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s.to_lowercase().as_str() {
            "transform" | "text2music" => Some(Self::Transform),
            "cover" => Some(Self::Cover),
            "repaint" => Some(Self::Repaint),
            "extract" | "separation" => Some(Self::Extract),
            "layer" | "lego" => Some(Self::Layer),
            "complete" | "accompaniment" => Some(Self::Complete),
            _ => None,
        }
    }

    pub fn description(&self) -> &'static str {
        match self {
            Self::Transform => "Generate music from text description",
            Self::Cover => "Create a cover version with different style",
            Self::Repaint => "Modify specific regions of audio",
            Self::Extract => "Separate audio sources (vocals, drums, etc.)",
            Self::Layer => "Add or remove instrument layers",
            Self::Complete => "Add accompaniment to existing audio",
        }
    }
}

/// Request to send to Nueva AI Bridge
#[derive(Debug, Serialize)]
struct BridgeRequest {
    mode: String,
    input_path: String,
    output_path: String,
    prompt: Option<String>,
    intensity: f32,
    preserve_melody: bool,
    preserve_tempo: bool,
    preserve_key: bool,
    extract_target: Option<String>,
    add_layers: Option<Vec<String>>,
    remove_layers: Option<Vec<String>>,
    duration_seconds: Option<f32>,
    seed: Option<i64>,
    quantization: Option<String>,
    extra_params: HashMap<String, serde_json::Value>,
}

/// Response from Nueva AI Bridge
#[derive(Debug, Deserialize)]
struct BridgeResponse {
    success: bool,
    output_path: Option<String>,
    processing_time_ms: u64,
    description: String,
    intentional_artifacts: Vec<String>,
    warnings: Vec<String>,
    error_code: Option<String>,
    error_message: Option<String>,
    #[serde(default)]
    metadata: HashMap<String, serde_json::Value>,
}

/// Real ACE-Step model implementation
pub struct AceStepModel {
    info: NeuralModelInfo,
    bridge_url: String,
    timeout_ms: u64,
    available: bool,
    gpu_info: Option<GpuInfo>,
    quantization: QuantizationLevel,
}

impl AceStepModel {
    /// Create a new ACE-Step model instance
    pub fn new() -> Self {
        let bridge_url =
            env::var("NUEVA_ACESTEP_API_URL").unwrap_or_else(|_| "http://localhost:8001".into());
        let timeout_ms = env::var("NUEVA_ACESTEP_TIMEOUT_MS")
            .ok()
            .and_then(|s| s.parse().ok())
            .unwrap_or(300_000); // 5 minutes default

        let gpu_info = GpuInfo::detect();
        let (available, quantization, _reason) = can_run_ace_step();

        Self {
            info: Self::create_model_info(),
            bridge_url,
            timeout_ms,
            available,
            gpu_info,
            quantization,
        }
    }

    /// Create a new instance with custom configuration
    pub fn with_config(bridge_url: String, timeout_ms: u64) -> Self {
        let gpu_info = GpuInfo::detect();
        let (available, quantization, _reason) = can_run_ace_step();

        Self {
            info: Self::create_model_info(),
            bridge_url,
            timeout_ms,
            available,
            gpu_info,
            quantization,
        }
    }

    fn create_model_info() -> NeuralModelInfo {
        NeuralModelInfo {
            id: "ace-step".to_string(),
            name: "ACE-Step 1.5".to_string(),
            version: "1.5".to_string(),
            description: "Full music transformation via Hybrid Reasoning-Diffusion".to_string(),
            capabilities: vec![
                "text_to_music".to_string(),
                "cover".to_string(),
                "repaint".to_string(),
                "style_change".to_string(),
                "track_extraction".to_string(),
                "layering".to_string(),
                "completion".to_string(),
            ],
            use_when: vec![
                "Dramatic transformation".to_string(),
                "genre change".to_string(),
                "cover generation".to_string(),
                "reimagine as X".to_string(),
                "separate tracks".to_string(),
                "add instruments".to_string(),
            ],
            limitations: vec![
                "Requires GPU for best performance".to_string(),
                "Non-deterministic output".to_string(),
                "Processing takes several seconds".to_string(),
            ],
            known_artifacts: vec![
                "Vocal intelligibility loss on complex lyrics".to_string(),
                "Tempo drift on pieces >5 minutes".to_string(),
                "Transient softening on aggressive percussion".to_string(),
            ],
            vram_requirement_gb: 4.0,
            inference_time: "1-30 seconds depending on GPU".to_string(),
            supported_params: vec![
                ParamSpec {
                    name: "mode".to_string(),
                    param_type: ParamType::Enum {
                        options: vec![
                            "transform".to_string(),
                            "cover".to_string(),
                            "repaint".to_string(),
                            "extract".to_string(),
                            "layer".to_string(),
                            "complete".to_string(),
                        ],
                    },
                    description: "Processing mode".to_string(),
                    default: Some(serde_json::json!("transform")),
                    required: false,
                },
                ParamSpec {
                    name: "prompt".to_string(),
                    param_type: ParamType::String,
                    description: "Text description of desired output".to_string(),
                    default: None,
                    required: true,
                },
                ParamSpec {
                    name: "intensity".to_string(),
                    param_type: ParamType::Float { min: 0.0, max: 1.0 },
                    description: "Transformation intensity (0-1)".to_string(),
                    default: Some(serde_json::json!(0.7)),
                    required: false,
                },
                ParamSpec {
                    name: "preserve_melody".to_string(),
                    param_type: ParamType::Bool,
                    description: "Preserve original melody (cover mode)".to_string(),
                    default: Some(serde_json::json!(true)),
                    required: false,
                },
                ParamSpec {
                    name: "preserve_tempo".to_string(),
                    param_type: ParamType::Bool,
                    description: "Preserve original tempo".to_string(),
                    default: Some(serde_json::json!(true)),
                    required: false,
                },
                ParamSpec {
                    name: "preserve_key".to_string(),
                    param_type: ParamType::Bool,
                    description: "Preserve original key".to_string(),
                    default: Some(serde_json::json!(true)),
                    required: false,
                },
                ParamSpec {
                    name: "extract_target".to_string(),
                    param_type: ParamType::Enum {
                        options: vec![
                            "vocals".to_string(),
                            "drums".to_string(),
                            "bass".to_string(),
                            "other".to_string(),
                            "all".to_string(),
                        ],
                    },
                    description: "What to extract (extract mode)".to_string(),
                    default: Some(serde_json::json!("vocals")),
                    required: false,
                },
                ParamSpec {
                    name: "seed".to_string(),
                    param_type: ParamType::Int {
                        min: -1,
                        max: i32::MAX,
                    },
                    description: "Random seed for reproducibility (-1 for random)".to_string(),
                    default: Some(serde_json::json!(-1)),
                    required: false,
                },
            ],
        }
    }

    /// Get GPU information
    pub fn gpu_info(&self) -> Option<&GpuInfo> {
        self.gpu_info.as_ref()
    }

    /// Get recommended quantization level
    pub fn quantization(&self) -> QuantizationLevel {
        self.quantization
    }

    /// Check if the bridge is reachable
    #[cfg(feature = "acestep")]
    fn check_bridge_health(&self) -> Result<bool> {
        let client = reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_secs(5))
            .build()
            .map_err(|e| NuevaError::BridgeConnectionError {
                message: e.to_string(),
            })?;

        let url = format!("{}/health", self.bridge_url);
        match client.get(&url).send() {
            Ok(resp) => Ok(resp.status().is_success()),
            Err(_) => Ok(false),
        }
    }

    #[cfg(not(feature = "acestep"))]
    fn check_bridge_health(&self) -> Result<bool> {
        Ok(false)
    }

    /// Send request to the bridge
    #[cfg(feature = "acestep")]
    fn send_request(&self, request: &BridgeRequest) -> Result<BridgeResponse> {
        let client = reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_millis(self.timeout_ms))
            .build()
            .map_err(|e| NuevaError::BridgeConnectionError {
                message: e.to_string(),
            })?;

        let url = format!("{}/process", self.bridge_url);

        let response = client
            .post(&url)
            .json(request)
            .send()
            .map_err(|e| {
                if e.is_timeout() {
                    NuevaError::AceStepTimeout {
                        timeout_ms: self.timeout_ms,
                    }
                } else if e.is_connect() {
                    NuevaError::AceStepUnavailable {
                        reason: format!("Cannot connect to bridge at {}: {}", self.bridge_url, e),
                    }
                } else {
                    NuevaError::BridgeConnectionError {
                        message: e.to_string(),
                    }
                }
            })?;

        if !response.status().is_success() {
            return Err(NuevaError::AceStepUnavailable {
                reason: format!("Bridge returned error: {}", response.status()),
            });
        }

        response.json::<BridgeResponse>().map_err(|e| {
            NuevaError::BridgeConnectionError {
                message: format!("Invalid response from bridge: {}", e),
            }
        })
    }

    #[cfg(not(feature = "acestep"))]
    fn send_request(&self, _request: &BridgeRequest) -> Result<BridgeResponse> {
        Err(NuevaError::AceStepUnavailable {
            reason: "ACE-Step support not compiled. Build with --features acestep".to_string(),
        })
    }
}

impl Default for AceStepModel {
    fn default() -> Self {
        Self::new()
    }
}

impl NeuralModel for AceStepModel {
    fn info(&self) -> &NeuralModelInfo {
        &self.info
    }

    fn is_available(&self) -> bool {
        self.available && self.check_bridge_health().unwrap_or(false)
    }

    fn validate_params(&self, params: &NeuralModelParams) -> Result<()> {
        // Check that prompt is provided
        if params.get_string("prompt").is_none() {
            return Err(NuevaError::InvalidParameter {
                param: "prompt".to_string(),
                value: "<missing>".to_string(),
                expected: "text description of desired output".to_string(),
            });
        }

        // Validate mode if provided
        if let Some(mode) = params.get_string("mode") {
            if AceStepMode::from_str(&mode).is_none() {
                return Err(NuevaError::InvalidParameter {
                    param: "mode".to_string(),
                    value: mode,
                    expected: "transform, cover, repaint, extract, layer, or complete".to_string(),
                });
            }
        }

        // Validate intensity range
        if let Some(intensity) = params.get_f32("intensity") {
            if !(0.0..=1.0).contains(&intensity) {
                return Err(NuevaError::InvalidParameter {
                    param: "intensity".to_string(),
                    value: intensity.to_string(),
                    expected: "value between 0.0 and 1.0".to_string(),
                });
            }
        }

        Ok(())
    }

    fn process(
        &self,
        input_path: &Path,
        output_path: &Path,
        params: &NeuralModelParams,
    ) -> Result<ProcessingResult> {
        let start = Instant::now();

        // Validate params
        self.validate_params(params)?;

        // Check VRAM if GPU info available
        if let Some(gpu) = &self.gpu_info {
            if !gpu.suitable_for_ace_step {
                tracing::warn!(
                    "GPU has insufficient VRAM ({:.1}GB), using CPU fallback",
                    gpu.vram_available_gb
                );
            }
        }

        // Extract parameters
        let mode = params
            .get_string("mode")
            .unwrap_or_else(|| "transform".to_string());
        let prompt = params.get_string("prompt");
        let intensity = params.get_f32("intensity").unwrap_or(0.7);
        let preserve_melody = params.get_bool("preserve_melody").unwrap_or(true);
        let preserve_tempo = params.get_bool("preserve_tempo").unwrap_or(true);
        let preserve_key = params.get_bool("preserve_key").unwrap_or(true);
        let extract_target = params.get_string("extract_target");
        let seed = params.get::<i64>("seed");

        // Build request
        let request = BridgeRequest {
            mode,
            input_path: input_path.to_string_lossy().to_string(),
            output_path: output_path.to_string_lossy().to_string(),
            prompt,
            intensity,
            preserve_melody,
            preserve_tempo,
            preserve_key,
            extract_target,
            add_layers: params.get("add_layers"),
            remove_layers: params.get("remove_layers"),
            duration_seconds: params.get_f32("duration"),
            seed,
            quantization: Some(match self.quantization {
                QuantizationLevel::FP32 => "fp32".to_string(),
                QuantizationLevel::FP16 => "fp16".to_string(),
                QuantizationLevel::INT8 => "int8".to_string(),
                QuantizationLevel::CPU => "cpu".to_string(),
            }),
            extra_params: HashMap::new(),
        };

        // Send to bridge
        let response = self.send_request(&request)?;

        let elapsed = start.elapsed().as_millis() as u64;

        if response.success {
            Ok(ProcessingResult {
                success: true,
                output_path: response.output_path,
                processing_time_ms: response.processing_time_ms.max(elapsed),
                description: response.description,
                intentional_artifacts: response.intentional_artifacts,
                warnings: response.warnings,
                metadata: response.metadata,
            })
        } else {
            Err(NuevaError::AiProcessingError {
                reason: response
                    .error_message
                    .unwrap_or_else(|| "Unknown ACE-Step error".to_string()),
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ace_step_mode_conversion() {
        assert_eq!(AceStepMode::from_str("cover"), Some(AceStepMode::Cover));
        assert_eq!(
            AceStepMode::from_str("extract"),
            Some(AceStepMode::Extract)
        );
        assert_eq!(AceStepMode::from_str("lego"), Some(AceStepMode::Layer));
        assert_eq!(
            AceStepMode::from_str("text2music"),
            Some(AceStepMode::Transform)
        );
        assert_eq!(AceStepMode::from_str("invalid"), None);
    }

    #[test]
    fn test_model_info() {
        let model = AceStepModel::new();
        let info = model.info();

        assert_eq!(info.id, "ace-step");
        assert_eq!(info.version, "1.5");
        assert!(info.vram_requirement_gb >= 4.0);
    }

    #[test]
    fn test_validate_params_missing_prompt() {
        let model = AceStepModel::new();
        let params = NeuralModelParams::new().with_param("mode", "cover");

        let result = model.validate_params(&params);
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_params_invalid_mode() {
        let model = AceStepModel::new();
        let params = NeuralModelParams::new()
            .with_param("prompt", "make it jazzy")
            .with_param("mode", "invalid_mode");

        let result = model.validate_params(&params);
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_params_valid() {
        let model = AceStepModel::new();
        let params = NeuralModelParams::new()
            .with_param("prompt", "convert to jazz style")
            .with_param("mode", "cover")
            .with_param("intensity", 0.8f32);

        let result = model.validate_params(&params);
        assert!(result.is_ok());
    }

    #[test]
    fn test_validate_params_invalid_intensity() {
        let model = AceStepModel::new();
        let params = NeuralModelParams::new()
            .with_param("prompt", "test")
            .with_param("intensity", 1.5f32);

        let result = model.validate_params(&params);
        assert!(result.is_err());
    }
}
